src/main/kotlin/uk/gov/justice/digital/hmpps/nomismappingservice/finance/PrisonerBalanceService.kt

```diff
@@ -0,0 +1,78 @@
+package uk.gov.justice.digital.hmpps.nomismappingservice.finance
+
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.flow.toList
+import org.springframework.data.domain.Page
+import org.springframework.data.domain.PageImpl
+import org.springframework.data.domain.Pageable
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.nomismappingservice.service.NotFoundException
+
+@Service
+@Transactional(readOnly = true)
+class PrisonerBalanceService(
+  private val repository: PrisonerBalanceMappingRepository,
+) {
+  suspend fun getMappingByNomisId(nomisRootOffenderId: Long) = repository.findOneByNomisRootOffenderId(nomisRootOffenderId)
+    ?.toDto()
+    ?: throw NotFoundException("No prisoner balance mapping found for nomisRootOffenderId=$nomisRootOffenderId")
+
+  suspend fun getMappingByDpsId(dpsId: String) = repository.findOneByDpsId(dpsId = dpsId)
+    ?.toDto()
+    ?: throw NotFoundException("No prisoner balance mapping found for dps Id=$dpsId")
+
+  suspend fun getMappingByDpsIdOrNull(dpsId: String) = repository.findOneByDpsId(dpsId = dpsId)
+    ?.toDto()
+
+  @Transactional
+  suspend fun createMapping(mapping: PrisonerBalanceMappingDto) {
+    repository.save(mapping.fromDto())
+  }
+
+  suspend fun getMappingsByMigrationId(pageRequest: Pageable, migrationId: String): Page<PrisonerBalanceMappingDto> = coroutineScope {
+    val mappings = async {
+      repository.findAllByLabelAndMappingTypeOrderByLabelDesc(
+        label = migrationId,
+        mappingType = PrisonerBalanceMappingType.MIGRATED,
+        pageRequest = pageRequest,
+      )
+    }
+
+    val count = async {
+      repository.countAllByLabelAndMappingType(
+        migrationId = migrationId,
+        mappingType = PrisonerBalanceMappingType.MIGRATED,
+      )
+    }
+    PageImpl(
```

Comment:
think we should be trying to avoid creating `PageImpl` these days and use `PagedModel` where possible